### ISSUE TITLE
static.File shouldn't allow excessive /s.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11100]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)| @jyknight reported|
|-|-|
|Trac ID|trac#1570|
|Type|defect|
|Created|2006-03-13 22:50:21Z|

Using ```static.File('/')```, I can successfully get the contents of /etc/resolv.conf with the url ```http://localhost:8080/etc//////resolv.conf////```. There's two things wrong with this:
 1. Empty segments. Either that should be an error or a redirect.
 2. a / after a file (not a directory) should surely be an error.

<details><summary>Searchable metadata</summary>

```
trac-id__1570 1570
type__defect defect
reporter__jknight jknight
priority__high high
milestone__Web2_Gold_Master Web2-Gold-Master
branch__ 
branch_author__ 
status__closed closed
resolution__wontfix wontfix
component__web2 web2
keywords__ 
time__1142290221000000 1142290221000000
changetime__1297607556000000 1297607556000000
version__ 
owner__ 
cc__wsanchez cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

I think that //// is tricky behavior to get right.  For example, http://apache.org/foundation////////contributing.html

A slash after a file should definitely be an error though.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_3"></a>|@wsanchez commented|
|-|-|

Given the security implications here, this should be a showstopper for a 1.0 web2 release.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_4"></a>|@jyknight commented|
|-|-|

Replying to [glyph](#note_2):
> I think that //// is tricky behavior to get right.  For example, http://apache.org/foundation////////contributing.html

That behavior falls under [#11338](https://github.com/twisted/twisted/issues/11338): it is incorrect and should not be supported.

**Comment by twisted-trac** [comment]
|<img alt="rikyu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|rikyu commented|
|-|-|

Wouldn't http://example.com/index.html/ be equivalent to http://example.com/index.html with a PATH_INFO of '/'? This is how apache handles it, although whether that means anything is left as an exercise to the reader.

I've used this approach often as a method of creating semi-clean URLs on webhosts that don't support mod_rewrite, as it's accessible from raw CGI, PHP, even client-side JavaScript. These may not be major concerns for web2, but if it gets us that closer to a web2 release, I think the trailing slash (and any subsequent path after the filename) should just be added to PATH_INFO.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

PATH_INFO is a CGI thing.  Environment variables probably aren't the answer here.  A slash after a file should definitely be an error.  I'm not sure why there's a question about two slashes in a row.  What's the motivation to translate /foo//bar into anything other than ```root.locateChild('foo').locateChild('').locateChild('bar')```?  The goal of the fundamental abstraction should be simplicity and consistency.  Given that, convenience can be built on top to make the nice thing simple, without compromising the system overall.

**Comment by twisted-trac** [comment]
|<img alt="rikyu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|rikyu commented|
|-|-|

Yeah, my mistake. I think I was in too much of a hurry to contribute in some way. ;-)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_6):
> I'm not sure why there's a question about two slashes in a row.

Apache does something weird there, and maybe it is worth doing the same weird thing for the sake of compatibility.  I'm not dead set on that, but compatibility is the reason I brought it up.

> What's the motivation to translate /foo//bar into anything other than root.locateChild('foo').locateChild('').locateChild('bar')?

Ignoring the obviously incorrect format of those API calls ;), I would never propose to treat this differently or specially in the "kernel" of the webserver.  If a compatibility hack is desirable, it's desirable specifically in ```static.File```, for compatibility with other webservers' file-serving semantics.

As long as this is a hot topic now, though, why is this so important that it's included in the gold master milestone?  It seems to be a small detail, which is arguably not a bug, with no security or performance implications, and few-to-no long-term support issues.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Replying to [wsanchez](#note_3):
> Given the security implications here, this should be a showstopper for a 1.0 web2 release.

Oops.  Since you said "there are security implications" and then I said "there aren't" let me clarify:

The example given in the ticket's description involves a static.File at "/", so while the paths _look_ scary, they aren't actually.  The problem being discussed isn't that you can access /etc/resolv.conf from an arbitrary file resource; it's that you can access /path/x/y/z with /x//////y///////z////// from a static file resource serving /path.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

Not going to fix this (See [#4821](https://github.com/twisted/twisted/issues/4821)).  Since `twisted.web.static.File` does not have the same problem, closing this.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
